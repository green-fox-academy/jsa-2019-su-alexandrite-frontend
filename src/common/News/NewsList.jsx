import React from 'react';
import {
  FlatList,
  View,
} from 'react-native';
import PropTypes from 'prop-types';
import NewsListItem from './NewsListItem';
import commonStyles from '../styles';

const NewsList = ({ news }) => (
  <FlatList
    data={news}
    keyExtractor={(item) => item.url}
    ItemSeparatorComponent={() => <View style={commonStyles.newsCardSeparator} />}
    renderItem={({ item }) => <NewsListItem item={item} />}
  />
);

NewsList.propTypes = {
  news: PropTypes.arrayOf(
    PropTypes.shape(NewsListItem.propTypes),
  ),
};

NewsList.defaultProps = {
  news: [],
};

export default NewsList;
