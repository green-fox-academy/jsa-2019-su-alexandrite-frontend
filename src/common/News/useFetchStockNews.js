import { useEffect, useState } from 'react';
import {
  NEWS_API_URL,
  NEWS_API_KEY,
} from 'react-native-dotenv';

const fetchNews = async (query, page, pageSize) => {
  const newsUrl = new URL(`${NEWS_API_URL}/everything`);
  const parsedQuery = (typeof query.join) === 'function'
    ? query.join(' OR ')
    : query;
  newsUrl.searchParams.append('apiKey', NEWS_API_KEY);
  newsUrl.searchParams.append('q', parsedQuery);
  newsUrl.searchParams.append('pageSize', pageSize);
  newsUrl.searchParams.append('page', page);
  newsUrl.searchParams.append('sortBy', 'publishedAt');
  newsUrl.searchParams.append('language', 'en');
  const res = await fetch(newsUrl);
  if (!res.ok) {
    switch (res.status) {
      case 404:
        throw new Error(`Cannot find any news about ${query}`);
      case 426:
        throw new Error(`No more news about ${query}, please come back later!`);
      default:
        throw new Error('Oops, there\'s something wrong with our app.');
    }
  }
  const json = await res.json();
  return json.articles.map(({
    publishedAt, title, urlToImage, url,
  }) => ({
    datetime: publishedAt,
    headline: title,
    image: urlToImage,
    url,
    id: (new Date() * Math.random()).toString(),
  }));
};

export default (query, page = 1, pageSize = 20) => {
  const [news, setNews] = useState([]);
  const [error, setError] = useState(null);
  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    setIsLoading(true);
    const fetchData = async () => {
      try {
        const result = await fetchNews(query, page, pageSize);
        setNews(page === 1 ? result : [...news, ...result]);
        setIsLoading(false);
      } catch (err) {
        setError(err);
        setIsLoading(false);
      }
    };
    fetchData();
  }, [page]);
  return {
    news, error, isLoading,
  };
};
