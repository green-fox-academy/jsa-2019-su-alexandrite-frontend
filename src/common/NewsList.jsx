import React from 'react';
import {
  FlatList,
  View,
} from 'react-native';
import PropTypes from 'prop-types';
import NewsCard from './NewsCard';
import commonStyles from './styles';

const NewsList = ({ news }) => (
  <FlatList
    data={news}
    keyExtractor={(item) => item.url}
    ItemSeparatorComponent={() => <View style={commonStyles.newsCardSeparator} />}
    renderItem={({ item }) => <NewsCard item={item} />}
  />
);

NewsList.propTypes = {
  news: PropTypes.arrayOf(
    PropTypes.shape(NewsCard.propTypes),
  ),
};

NewsList.defaultProps = {
  news: [],
};


export default NewsList;
