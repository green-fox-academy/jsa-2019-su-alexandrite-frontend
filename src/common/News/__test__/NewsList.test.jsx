import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { useDispatch } from 'react-redux';
import { useNavigation } from 'react-navigation-hooks';
import useFetchStockNews from '../useFetchStockNews';
import NewsList, { ListFooter } from '../NewsList';

jest.mock('react-redux');
jest.mock('react-navigation-hooks');
jest.mock('../useFetchStockNews');
jest.mock('../NewsMappedList', () => 'NewsMappedList');
jest.mock('../NewsListItem', () => 'NewsListItem');

describe('<NewsList />', () => {
  beforeEach(() => {
    useDispatch.mockReturnValue(jest.fn());
    useFetchStockNews.mockReturnValue({
      news: [],
      isLoading: false,
      error: null,
    });
    useNavigation.mockReturnValue({
      push: jest.fn(),
      navigate: jest.fn(),
    });
  });

  afterAll(() => {
    jest.resetModules();
  });

  it('Should render properly and match the snapshot', () => {
    const tree = renderer.create(
      <NewsList query="test" />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should render properly and match the snapshot on error', () => {
    useFetchStockNews.mockReturnValueOnce({
      news: null,
      isLoading: false,
      error: {
        message: 'test error',
      },
    });
    const tree = renderer.create(
      <NewsList query="test" />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should render properly and match the snapshot when loading', () => {
    useFetchStockNews.mockReturnValueOnce({
      news: null,
      isLoading: true,
      error: null,
    });
    const tree = renderer.create(
      <NewsList query="test" />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('keyExtractor should extract the id field of the items', () => {
    const wrapper = shallow(
      <NewsList query="test" />,
    );
    expect(wrapper.find('FlatList').props().keyExtractor({ id: '123' })).toBe('123');
  });

  it('ItemSeparatorComponent should match snapshot', () => {
    const wrapper = shallow(
      <NewsList query="test" />,
    );
    const separator = renderer.create(
      wrapper.find('FlatList').props().ItemSeparatorComponent(),
    ).toJSON();
    expect(separator).toMatchSnapshot();
  });

  it('renderItem should render NewsListItem and match snapshot', () => {
    const wrapper = shallow(
      <NewsList query="test" />,
    );
    const item = renderer.create(
      wrapper.find('FlatList').props().renderItem({ item: {} }),
    ).toJSON();
    expect(item).toMatchSnapshot();
  });

  it('listFooter should render and match snapshot on Loading', () => {
    const footer = renderer.create(
      <ListFooter errorMessage="" isLoading pageNumber={2} />,
    ).toJSON();
    expect(footer).toMatchSnapshot();
  });


  it('listFooter should render and match snapshot on onError', () => {
    const footer = renderer.create(
      <ListFooter errorMessage="error" isLoading={false} pageNumber={2} />,
    ).toJSON();
    expect(footer).toMatchSnapshot();
  });

  it('listFooter should not render when no error and not loading', () => {
    const footer = renderer.create(
      <ListFooter errorMessage="" isLoading={false} pageNumber={2} />,
    ).toJSON();
    expect(footer).toMatchSnapshot();
  });

  it('listFooter should not render when pageNumber is 1', () => {
    expect(ListFooter({ pageNumber: 1 })).toBeNull();
  });
});
