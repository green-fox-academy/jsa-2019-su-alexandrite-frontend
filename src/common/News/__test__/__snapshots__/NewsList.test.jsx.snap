// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewsList /> ItemSeparatorComponent should match snapshot 1`] = `
<View
  style={
    Object {
      "height": 15,
    }
  }
/>
`;

exports[`<NewsList /> Should render properly and match the snapshot 1`] = `
<RCTScrollView
  ItemSeparatorComponent={[Function]}
  ListFooterComponent={
    <ListFooter
      errorMessage={null}
      isLoading={false}
      pageNumber={1}
    />
  }
  ListHeaderComponent={[Function]}
  contentContainerStyle={
    Object {
      "padding": 15,
    }
  }
  data={Array []}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReached={[Function]}
  onEndReachedThreshold={0.9}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onRefresh={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={false}
    />
  }
  refreshing={false}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={
    Array [
      0,
    ]
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <RCTRefreshControl />
  <View>
    <View
      onLayout={[Function]}
    />
    <View
      onLayout={[Function]}
    />
  </View>
</RCTScrollView>
`;

exports[`<NewsList /> Should render properly and match the snapshot on error 1`] = `
<RCTScrollView
  ItemSeparatorComponent={[Function]}
  ListFooterComponent={
    <ListFooter
      errorMessage="test error"
      isLoading={false}
      pageNumber={1}
    />
  }
  ListHeaderComponent={[Function]}
  contentContainerStyle={
    Object {
      "padding": 15,
    }
  }
  data={null}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReached={[Function]}
  onEndReachedThreshold={0.9}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onRefresh={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={false}
    />
  }
  refreshing={false}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={
    Array [
      0,
    ]
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <RCTRefreshControl />
  <View>
    <View
      onLayout={[Function]}
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
              "borderRadius": 6,
              "elevation": 3,
              "marginBottom": 15,
              "padding": 15,
              "shadowColor": "#000",
              "shadowOffset": Object {
                "x": 0,
                "y": 4,
              },
              "shadowOpacity": 0.1,
              "shadowRadius": 6,
            },
            null,
          ]
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#fff6f6",
              "borderRadius": 6,
              "paddingHorizontal": 8,
              "paddingVertical": 15,
              "width": 690,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#9f3a38",
                "fontWeight": "bold",
              }
            }
          >
            test error
          </Text>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
    />
  </View>
</RCTScrollView>
`;

exports[`<NewsList /> Should render properly and match the snapshot when loading 1`] = `
<RCTScrollView
  ItemSeparatorComponent={[Function]}
  ListFooterComponent={
    <ListFooter
      errorMessage={null}
      isLoading={true}
      pageNumber={1}
    />
  }
  ListHeaderComponent={[Function]}
  contentContainerStyle={
    Object {
      "padding": 15,
    }
  }
  data={null}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReached={[Function]}
  onEndReachedThreshold={0.9}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onRefresh={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={true}
    />
  }
  refreshing={true}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={
    Array [
      0,
    ]
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <RCTRefreshControl />
  <View>
    <View
      onLayout={[Function]}
    />
    <View
      onLayout={[Function]}
    />
  </View>
</RCTScrollView>
`;

exports[`<NewsList /> listFooter should not render when no error and not loading 1`] = `null`;

exports[`<NewsList /> listFooter should render and match snapshot on Loading 1`] = `
<View
  style={
    Object {
      "padding": 15,
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="small"
  />
</View>
`;

exports[`<NewsList /> listFooter should render and match snapshot on onError 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "borderRadius": 6,
        "elevation": 3,
        "marginBottom": 15,
        "padding": 15,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "x": 0,
          "y": 4,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 6,
      },
      Object {
        "marginTop": 15,
      },
    ]
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#fff6f6",
        "borderRadius": 6,
        "paddingHorizontal": 8,
        "paddingVertical": 15,
        "width": 690,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#9f3a38",
          "fontWeight": "bold",
        }
      }
    >
      error
    </Text>
  </View>
</View>
`;

exports[`<NewsList /> renderItem should render NewsListItem and match snapshot 1`] = `
<NewsListItem
  item={Object {}}
/>
`;
